### ISSUE TITLE
document web form and command-line options

### ISSUE THREAD

**Issue by VladimirAlexiev** [gkellogg/rdf-distiller#16]
Some of the options on the web form are non-obvious, and I can't find them in the documentation http://rdf.greggkellogg.net/yard/RDF/Distiller/Application.html (probably because the search is not full-text search but only class/method names). Some of the options have an informative tooltip, but it's not always enough, and others don't have it.
- `expandContext`: I guess this is context used for expanding? Why do we need a second context for that? ("flatten" and "compact" don't get such option, "expand" and "frame" do). Cant's see it in http://rdf.greggkellogg.net/yard/JSON/LD/Expand.html#expand-instance_method
- `rdfDirection`: please explain this
- Where are command-line options documented?

The following come from https://www.w3.org/TR/json-ld11-framing/#framing-keywords. But it'd be very useful if you describe them and link to motivating examples in the spec 
- `explicit` "Only include explicitly declared properties in output"
- `omitDefault` "Omit missing properties from output": what else could you do with missing properties, output them as `null`? 
- `requireAll` "Require all properties to match": (the further mention of `--no-compact-to-relative` is wrong)
- `embed`: please explain this

If this is already documented, please excuse me and point me to the doc. Cheers!

**Comment by gkellogg** [comment]
These options refer to JSON-LD specific options, defined in the [JSON-LD API](https://www.w3.org/TR/json-ld11-api/#the-jsonldoptions-type) and [JSON-LD Framing](https://www.w3.org/TR/json-ld11-framing/#jsonldoptions) and are only available when the input format is "jsonld".

For more information see those documents and the [JSON-LD Syntax](https://www.w3.org/TR/json-ld11/).

The Distiller form elements are derived from the command line options of the `rdf` CLI, which can be seen with `--help` (after selecting an appropriate input- or output- format). In turn, these are defined in either the Format, Reader or Writer classes. (E.g., [JSON::LD::Format](https://github.com/ruby-rdf/json-ld/blob/241ef57e096cefdff55ef6d39a917049eec04529/lib/json/ld/format.rb#L50-L214), [JSON::LD::Reader](https://github.com/ruby-rdf/json-ld/blob/241ef57e096cefdff55ef6d39a917049eec04529/lib/json/ld/reader.rb#L15-L50), or [JSON::LD::Writer](https://github.com/ruby-rdf/json-ld/blob/241ef57e096cefdff55ef6d39a917049eec04529/lib/json/ld/writer.rb#L75-L167).

Examples in those files could certainly be improved, with links into examples in the specs.

* `expandContext` is there to provide an external link to a context to use when expanding a JSON file that should be considered as JSON-LD. There is also a way to do this with a Link header.
* `rdfDirection` is a new feature added to handle the `@direction` keyword added to JSON-LD. Eventually, we expect some text direction mechanism to be added to RDF. For now, this provides for two non-normative mechanisms using a datatype or a compound literal. You can read up on this in the syntax document.
* `explicit` eliminates from the output any properties not described explicitly in the framing document.
* `omitDefault` overrides any `@default` included in the frame to not include that in the output.
* `requireAll` is used for frame matching to require that all properties described in a sub-frame be present to match (or not). (I'll fix the documentation to use `--no-requireAll` instead).
* `embed` controls if referenced objects are embedded, or described elsewhere. It can be useful to have objects always be present, even if it creates duplicates, for consistency, or to force them out to the outer-most level.

Please consider helping to improve documentation in the various gems with PRs.

**Comment by VladimirAlexiev** [comment]
Immense thanks!

In particular, `expandContext` is very useful and I think is missing at the jsonld playground. 
But I still don't understand what's the difference from `context` and why two of the commands need 2 contexts.

To add PRs, I think 2 UI  improvements are needed:
- ability to use hyperlinks in the explanations 
- hover displays a transient tooltip, but the user can't click inside it.
  - add an (i) icon: when clicked, display a popup (so then the user can click a link inside it). Dismiss when clicked a second time

**Comment by gkellogg** [comment]
Yes, the json-ld playground doesn't expose all options, but it's lacking adequate developer support. My UI is built automatically based on options specifically defined for readers and writers.

The "expandContext" is used for expanding a document, while the "context" argument is used when compacting. It's conceivable that you need to apply a context when expanding a native JSON document, and then want to compact (or frame) it with a different context.

I don't have much time to work on it now, but again, feel free to clone and submit PRs. I think having an (i) icon to bring up more documentation would be great. Ideally, it would somehow take you to the appropriate Yard page for a given option, but more likely would need to repeat the help content in the option description with Markdown. The Markdown could be processed on the server side and sent across in the JSON that drives the UI.

For example, the Yard documentation for the "expandContext" option is buried in http://rdf.greggkellogg.net/yard/JSON/LD/API.html#class_method_details (it should be in http://rdf.greggkellogg.net/yard/JSON/LD/Reader.html#initialize-instance_method, but I need to look at the tags. Unfortunately, there's no good way to introduce fragment identifiers in the generated HTML.

This documentation could be improved to reference the [definition on JSON-LD API](https://www.w3.org/TR/json-ld11-api/#dom-jsonldoptions-expandcontext), although that doesn't bring you to an adequate example, either. For that you'd need to see [example 142 in JSON-LD](https://www.w3.org/TR/json-ld11/#example-142-referencing-a-json-ld-context-from-a-json-document-via-an-http-link-header), and even then this uses the equivalent Link header.
